src/vision/status_quo/barbara_plays_with_async.md

```diff
@@ -0,0 +1,387 @@
+docs and finds the `join_all` function which looks like what she wants:
+
+```rust
+let mut futures = Vec::new();
+for _ in 0..10 {
+    let delay = rng.sample(t);
+    futures.push(tokio::time::sleep(Duration::from_millis(delay)).then(|_| {
+        println!("Done!");
+        std::future::ready(())
+    }));
+}
+println!("Created 10 futures");
+
+futures::future::join_all(futures).await;
+println!("Done");
+```
+It works exactly as expected now!  After having written the code, Barbara begins
+to remember an important detail about rust futures that she once read somewhere:
+rust futures are lazy, and won't make progress unless you await them.
+
+Happy with this success, Barbara continues to expand her toy program by making
+a few small adjustments:
+
+```rust
+for counter in 0..10 {
+    let delay = rng.sample(t);
+    let delay_future = tokio::time::sleep(Duration::from_millis(delay));
+
+    if counter < 9 {
+        futures.push(delay_future.then(|_| {
+            println!("Done!");
+            std::future::ready(())
+        }));
+    } else {
+        futures.push(delay_future.then(|_| {
+            println!("Done with the last future!");
+            std::future::ready(())
+        }));
+    }
+}
+```
+
+This fails to compile:
+
+```
+error[E0308]: mismatched types
+
+   = note: expected closure `[closure@src\main.rs:16:44: 19:14]`
+              found closure `[closure@src\main.rs:21:44: 24:14]`
+   = note: no two closures, even if identical, have the same type
+   = help: consider boxing your closure and/or using it as a trait object
+```
+
+This error doesn't actually surprise Barbara that much, as she is familiar with
+the idea of having to box objects sometimes.  She does
+notice the "consider boxing your closure" error, but thinks that this is not
+likely the correct solution.  Instead, she thinks that she should box the
+entire future.
+
+She first adds explicit type annotations to the Vec:
+
+```rust
+let mut futures: Vec<Box<dyn Future<Output=()>>> = Vec::new();
+```
+
+She then notices that her IDE (VSCode + rust-analyzer) has a new error on
+each call to push.  The code assist on each error says `Store this in the heap
+by calling 'Box::new'`.  She is exactly what she wants, and it happy that
+rust-analzyer perfectly handled this case.
+
+Now each future is boxed up, but there is one final error still,
+this time on the call to `join_all(futures).await`:
+
+```
+error[E0277]: `dyn futures::Future<Output = ()>` cannot be unpinned
+  --> src\main.rs:34:31
+   |
+34 |     futures::future::join_all(futures).await;
+```
+
+Barbara has been around rust for long enough to know that there is a `Box::pin`
+API, but she doesn't really understand what it does, nor does she have a good
+intuition about what this API is for.  But she is accustomed to just trying
+things in rust to see if they work.  And indeed, after changing `Box::new` to
+`Box::pin` (and adjusting the type of the Vec), the code compiles and runs
+successfully.
```

Comment:
Code sample?

Reply:
I pushed a commit that shows how each part changed.  In order to avoid repeating basically the same block of code over-and-over, I just showed the changed parts.  But if you think it'll read better to show the whole thing, let me know.